### PR TITLE
Integrate Agroal Datasource pool #1157

### PR DESCRIPTION
**Pull request by johnaohara**
Integrate Agroal database connection pool to provide a fast, efficient database connection pool when a full jca compliant database connection pool is not required.

### CODE REVIEW

File: core/agroal/src/main/java/io/thorntail/agroal/AgroalPoolMetaData.java [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+     * @param validateConnection The initial pool size.
+     * @return This meta-data object.
+     */
+    public AgroalPoolMetaData setValidateConnection(boolean validateConnection) {
+        this.validateConnection = validateConnection;
+        return this;
+    }
+
+    /**
+     * Are transactions are managed by JTA.
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/AgroalPoolMetaData.java:323]: Remove the second "are".

File: core/agroal/src/main/java/io/thorntail/agroal/AgroalPoolMetaData.java [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+     *
+     * @param xa are database connections are XA compliant.
+     * @return This meta-data object.
+     */
+    public AgroalPoolMetaData setXa(boolean xa) {
+        this.xa = xa;
+        return this;
+
+    }
+    public boolean isConnectable() {
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/AgroalPoolMetaData.java:362]: What does this attribute mean?

File: core/agroal/src/main/java/io/thorntail/agroal/AgroalPoolMetaData.java [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+
+    private int maxSize;
+
+    private Duration leakTimeout = ZERO;
+
+    private Duration validationTimeout = ZERO;
+
+    private Duration reapTimeout = ZERO;
+
+    private volatile Duration acquisitionTimeout = ZERO;
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/AgroalPoolMetaData.java:410]: Wonder why just 2 fields are `volatile`?

File: core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalMessages.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    @LogMessage(level = Logger.Level.WARN)
+    @Message(id = 2 + OFFSET, value = "no registered JDBC drivers")
+    void noRegisteredJDBCdrivers();
+
+    @LogMessage(level = Logger.Level.ERROR)
+    @Message(id = 3 + OFFSET, value = "JDBC driver class not found %s")
+    void noJDBCdriverClass(String className);
+
+    @LogMessage(level = Logger.Level.WARN)
+    @Message(id = 4 + OFFSET, value = "agroal datasource '%s' specified no JDBC driver, too many to choose from")
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalMessages.java:39]: What does "no JDBC driver, too many to choose from" mean?

File: core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalMessages.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    void noDriverSpecifiedManyDrivers(String datasource);
+
+    @LogMessage(level = Logger.Level.WARN)
+    @Message(id = 5 + OFFSET, value = "agroal datasource '%s' specified no JDBC driver, using %s")
+    void implicitlyUsingDriver(String datasource, String driver);
+
+    @LogMessage(level = Logger.Level.WARN)
+    @Message(id = 6 + OFFSET, value = "agroal datasource '%s' specified requested tracing, but tracing is not available")
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalMessages.java:47]: Remove "specified"?

File: core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalProducer.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            AgroalMessages.MESSAGES.noRegisteredJDBCdrivers();
+            return null;
+        }
+
+        AgroalDataSource dataSource = getDatasource();
+
+        try {
+            jndi.bind(metaData.getJNDIName(), dataSource);
+        } catch (NamingException e) {
+            e.printStackTrace();
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalProducer.java:74]: Hmm...

File: core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalProducer.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    public AgroalDataSource getDatasource() throws SQLException {
+        Class<?> providerClass = null;
+        try {
+            providerClass = this.getClass().getClassLoader().loadClass(jdbcDriverRegistry.get(this.metaData.getDriver()).getDriverClassName());
+        } catch (ClassNotFoundException e) {
+            AgroalMessages.MESSAGES.noJDBCdriverClass(jdbcDriverRegistry.get(this.metaData.getDriver()).getDriverClassName());
+        }
+        if (this.metaData.isXa()) {
+            if (!XADataSource.class.isAssignableFrom(providerClass)) {
+                throw new RuntimeException("Driver is not an XA datasource and xa has been configured");
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalProducer.java:90]: Shouldn't we use the JBoss Logging error mechanism uniformly? Few lines above, we use `AgroalMessages.MESSAGES.noJDBCdriverClass` to throw an exception (at least I hope it throws an exception :-) ), but not here.

File: core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalProducer.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+            }
+        } else {
+            if (providerClass != null && !DataSource.class.isAssignableFrom(providerClass) && !Driver.class.isAssignableFrom(providerClass)) {
+                throw new RuntimeException("Driver is an XA datasource and xa has been configured");
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/impl/AgroalProducer.java:94]: The error message doesn't make a lot of sense.

File: core/agroal/src/main/java/io/thorntail/agroal/impl/opentracing/TracedDataSource.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public boolean isWrapperFor(Class<?> iface) throws SQLException {
+        return delegate.isWrapperFor(iface);
+    }
+
+
+    private final AgroalDataSource delegate;
+//    private final TracedLocalManagedConnectionFactory mcf;
```
Comment (Ladicek) [inline core/agroal/src/main/java/io/thorntail/agroal/impl/opentracing/TracedDataSource.java:86]: This seems unfinished?

File: core/jpa-agroal/src/main/java/io/thorntail/jpa/JpaFraction.java [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package io.thorntail.jpa;
+
+/**
+ * @author kg6zvp
+ */
+public class JpaFraction { }
```
Comment (Ladicek) [inline core/jpa-agroal/src/main/java/io/thorntail/jpa/JpaFraction.java:6]: Is this totally useless? Thorntail v4 has no concept of fractions.

**Comment by juangon** [comment]
@johnaohara please can you check the review comments about this PR?

Thanks!

**Comment by johnaohara** [comment]
@juangon  @Ladicek  Only just seen the comments, missed the notifications. I will address the comments asap. Thanks

**Review by kenfinnigan — APPROVED** [review]

**Pull request merged**